src/proto/h2/mod.rs

```diff
@@ -13,13 +13,33 @@ pub(crate) use self::client::Client;
 pub(crate) use self::server::Server;
 
 fn strip_connection_headers(headers: &mut HeaderMap) {
-    if headers.remove(TRANSFER_ENCODING).is_some() {
-        trace!("removed illegal Transfer-Encoding header");
-    }
-    if headers.contains_key(CONNECTION) {
-        warn!("Connection header illegal in HTTP/2");
-        //TODO: actually remove it, after checking the value
-        //and removing all related headers
+    let connection_headers = [
+        CONNECTION,
+        TRANSFER_ENCODING,
+        UPGRADE,
+        HeaderName::from_lowercase(b"keep-alive").unwrap(),
+        HeaderName::from_lowercase(b"proxy-connection").unwrap(),
+    ];
+
+    for conn_header in connection_headers.iter() {
+        if let Some(header) = headers.remove(conn_header) {
+            warn!(
+                "Connection header illegal in HTTP/2: {}",
+                conn_header.as_str()
+            );
+            let header_contents = header.to_str().unwrap();
+
+            // A `Connection` header may have a comma-separated list of names of other headers that
+            // are meant for only this specific connection.
+            //
+            // Iterate these names and remove them as headers. Connection-specific headers are
+            // forbidden in HTTP2, as that information has been moved into frame types of the h2
+            // protocol.
+            for name in header_contents.split(',') {
```

Comment:
Actually it might be worth treating `Connection` separately, to avoid this.

Reply:
I agree, this part should only be done for the `CONNECTION` header. If someone had some weird custom `chunked: roflcopters` header, then it could be removed because of removing `Transfer-Encoding: chunked`.

Reply:
Updated to remove the comma separated headers listed only in the `CONNECTION` header.